fs: extends recursive readdir to allow a function to limit traversal (#49255)
nodejs/node: doc/api/fs.md

* `options` {string|Object}
  * `encoding` {string} **Default:** `'utf8'`
  * `withFileTypes` {boolean} **Default:** `false`
  * `recursive` {boolean|Function} **Default:** `false`

>     * `name` {string} A given directory file system artifact name.

Reviewer: ```suggestion
    * `name` {string|fs.Dirent} A given directory file system artifact name or the `Dirent` instance associated with it.
```

>     * `name` {string} A given directory file system artifact name.

Reviewer 2: Or also it can be `Buffer`, depending on `encoding` option
```suggestion
    * `name` {string|Buffer|fs.Dirent} A given directory file system artifact name or the `Dirent` instance associated with it.
```

  * `encoding` {string} **Default:** `'utf8'`
  * `withFileTypes` {boolean} **Default:** `false`
  * `recursive` {boolean} **Default:** `false`

>   * `depth` {number} **Default:** `1`

Reviewer: ```suggestion
  * `depth` {integer} **Default:** `1`
```

If `options.withFileTypes` is set to `true`, the resolved array will contain
{fs.Dirent} objects.

`options.depth` only applies if option `options.recursive` is set to `true` and
determines how many descendant depths of file system artifacts to include. A

> value of 0 eliminates recursion while a value of 1 traverses immediately child

Reviewer 3: ```suggestion
value of 0 eliminates recursion while a value of 1 traverses immediate child
```

If `options.withFileTypes` is set to `true`, the resolved array will contain
{fs.Dirent} objects.

> The `options.depth` determines the amount of recursion to apply. A value less
> than 1 applies full recursion, value of 1 applies no recursion, and greater

Reviewer: ```suggestion
The `options.depth` determines the amount of recursion to apply. A value of
`Infinity` applies full recursion, value of 1 applies no recursion, and greater
```

values determine the depth of descendant directies to traverse. If
`options.recursive` is value `true` then `options.depth` defaults to a value

> of `0` and otherwise defaults to a value `1`.

Reviewer: ```suggestion
of `Infinity` and otherwise defaults to a value `1`.
```

If `options.withFileTypes` is set to `true`, the `files` array will contain
{fs.Dirent} objects.

`options.depth` only applies if option `options.recursive` is set to `true` and
determines how many descendant depths of file system artifacts to include. A

> value of 0 eliminates recursion while a value of 1 traverses immediately child

Reviewer 3: ```suggestion
value of 0 eliminates recursion while a value of 1 traverses immediate child
```

If `options.withFileTypes` is set to `true`, the result will contain
{fs.Dirent} objects.

`options.depth` only applies if option `options.recursive` is set to `true` and
determines how many descendant depths of file system artifacts to include. A

> value of 0 eliminates recursion while a value of 1 traverses immediately child

Reviewer 3: ```suggestion
value of 0 eliminates recursion while a value of 1 traverses immediate child
```

The base path that this {fs.Dirent} object refers to.

### Class: `fs.FilterOutput`

A {fs.FilterOutput} object provides identifiers for file system artifacts. This
object is passed into the function provided to option `recursiveFilter` of
[`fsPromises.readdir()`][], [`fs.readdir()`][], and [`fs.readdirSync()`][] when
option `recursive` is assigned `true`.

```console
FilterOutput {
  depth: 2,
  itemName: 'fileName.ts',
  itemPath: '/users/userAccount/notes/fileName.ts',
  parentName: 'notes',
  parentPath: '/users/userAccount/notes'
}

> ```

Reviewer 3: Most of this seems pretty redundant. For consistency with `fs.cp()`, why not just pass the (potentially relative) path to the function? Determining the depth is simple enough if necessary, and so is determining the `parentName` etc.
